### PR TITLE
Provide more suggestions on invalid equality where bounds

### LINKED ISSUE #68982: Suggest correct code for associated fn with type parameters
#68689 partially addresses #50734, but we need to resugar the predicates to unify the output of a trait obligation with related projection obligations (associated types).

For example

```
   = help: implement the missing item: `fn from_iter<T>(_: T) -> Self where T: std::iter::IntoIterator, std::iter::IntoIterator::Item = A { unimplemented!() }`
```

should be

```
   = help: implement the missing item: `fn from_iter<T>(_: T) -> Self where T: std::iter::IntoIterator<Item = A> { unimplemented!() }`
```

### PR DESCRIPTION
**Pull request by estebank**
```
error: equality constraints are not yet supported in `where` clauses
  --> $DIR/equality-bound.rs:50:9
   |
LL |         IntoIterator::Item = A
   |         ^^^^^^^^^^^^^^^^^^^^^^ not supported
   |
   = note: see issue #20041 <https://github.com/rust-lang/rust/issues/20041> for more information
help: if `IntoIterator::Item` is an associated type you're trying to set, use the associated type binding syntax
   |
LL ~     fn from_iter<T: IntoIterator<Item = A>>(_: T) -> Self
LL ~
   |

error: equality constraints are not yet supported in `where` clauses
  --> $DIR/equality-bound.rs:63:9
   |
LL |         T::Item = A
   |         ^^^^^^^^^^^ not supported
   |
   = note: see issue #20041 <https://github.com/rust-lang/rust/issues/20041> for more information
help: if `IntoIterator::Item` is an associated type you're trying to set, use the associated type binding syntax
   |
LL ~     fn from_iter<T: IntoIterator<Item = A>>(_: T) -> Self
LL ~
   |
```

Fix #68982.

### CODE REVIEW

File: tests/ui/generic-associated-types/equality-bound.stderr [hunk 1]
```diff
@@ -32,6 +32,96 @@ LL | fn sum3<J: Iterator>(i: J) -> i32 where I::Item = i32 {
...
+  --> $DIR/equality-bound.rs:23:9
+   |
+LL |         IntoIterator::Item = A,
+   |         ^^^^^^^^^^^^^^^^^^^^^^ not supported
+   |
+   = note: see issue #20041 <https://github.com/rust-lang/rust/issues/20041> for more information
+help: if `IntoIterator::Item` is an associated type you're trying to set, use the associated type binding syntax
+   |
+LL ~         T: IntoIterator<Item = A>,
+LL ~         ,
```
Comment (compiler-errors) [inline tests/ui/generic-associated-types/equality-bound.stderr:45]: This doesn't look right

Comment (estebank) [reply tests/ui/generic-associated-types/equality-bound.stderr:45]: That's a preexisting issue with the suggestion machinery that I'll clean up.

File: compiler/rustc_ast_passes/src/ast_validation.rs [hunk 2]
```diff
@@ -1593,6 +1593,58 @@ fn deny_equality_constraints(
             }
         }
     }
+
+    let mut suggest =
```
Comment (nnethercote) [inline compiler/rustc_ast_passes/src/ast_validation.rs:1597]: +1 for factoring out the common code.

**Comment by nnethercote** [comment]
Looks ok to me, judging more on the changes to the tests outputs than the compiler code itself. So r=me but I'm not sure if @compiler-errors wants to weigh in again.

**Pull request merged**